68 add injection initializer

Hubert-Guzowski opened the pull request:
Changes introduced in these commits do not alter the functionality of HMS, instead making the interface more flexible and moving some of the logic away from demes.

WojtAcht commented on pyhms/core/initializers.py:
```diff
@@ -0,0 +1,120 @@
+
+    def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
+        sample = self.sampler.random(pop_size)
+        genomes = self.lower_bounds + sample * (self.upper_bounds - self.lower_bounds)
```
We should extract this to a function.
```
def linear_scale(x: np.ndarray, bounds: np.ndarray) -> np.ndarray:
  return bounds[:, 0] + sample * (bounds[:, 1] - bounds[:, 0])
```

WojtAcht replied:
and then we don't need:
```
self.lower_bounds = bounds[:, 0]
self.upper_bounds = bounds[:, 1]
```

WojtAcht commented on pyhms/core/initializers.py:
```diff
@@ -0,0 +1,120 @@
+    def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
+        sample = self.sampler.random(pop_size)
+        genomes = self.lower_bounds + sample * (self.upper_bounds - self.lower_bounds)
+        return [Individual(genome, problem=problem) for genome in genomes]
```
maybe we can have a method in `PopInitializer`:
```
def sample_to_individuals(self, sample: np.ndarray, problem: Problem) -> list[Individual]:
        return [Individual(genome=genome, problem=problem) for genome in sample]
```
and then we have:
```
class SobolGlobalInitializer(PopInitializer):
     def __init__(self, bounds: np.ndarray, random_seed: int = None):
        super().__init__(bounds)
        self.sampler = Sobol(d=len(bounds), scramble=True, seed=random_seed)

    def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
        sample = self.sampler.random(pop_size)
        return self.sample_to_individuals(linear_scale(sample, problem.bounds), problem)
```

WojtAcht replied:
and maybe all samplers can work like this? to just initialize the sampler and then you have the common sample_pop method:
```
def sample_pop(self, pop_size: int, problem: Problem) -> list[Individual]:
        sample = self.sampler.random(pop_size)
        return self.sample_to_individuals(linear_scale(sample, problem.bounds), problem)
```

WojtAcht commented on pyhms/core/initializers.py:
```diff
@@ -0,0 +1,120 @@
+        pass
+
+    def __call__(self, pop_size: int | None, problem: Problem):
+        return self.sample_pop(pop_size, problem)
```
why not just?
```

class PopInitializer(ABC):
    def __init__(self, bounds: np.ndarray | None = None):
        self._bounds = bounds

    @abstractmethod
    def __call__(self, pop_size: int | None, problem: Problem):
        return self.sample_pop(pop_size, problem)
```

WojtAcht commented on pyhms/demes/random_deme.py:
```diff
@@ -0,0 +1,37 @@
+        if (gsc_value := tree._gsc(tree)) or self._lsc(self):
+            self._active = False
+            message = (
+                "Random sampler Deme finished due to GSC" if gsc_value else "Random sampler Deme finished due to LSC"
```
We could also log `self._sampler.__class__.__name__` here

WojtAcht commented on pyhms/demes/initialize.py:
```diff
@@ -3,61 +3,90 @@
     random_seed: int = None,
     parent_deme: AbstractDeme | None = None,
 ) -> AbstractDeme:
+    child_initializer: PopInitializer
+    if config.pop_initializer_class == UniformGlobalInitializer:
+        child_initializer = UniformGlobalInitializer(config.bounds)
+    elif config.pop_initializer_class == GaussianInitializer:
+        if hasattr(config, "sample_std_dev"):
+            sample_std_dev = config.sample_std_dev
+        else:
+            sample_std_dev = get_default_mutation_std(config.bounds, target_level)
+        child_initializer = GaussianInitializer(seed=sprout_seed.genome, std_dev=sample_std_dev, bounds=config.bounds)
+    elif config.pop_initializer_class == GaussianInitializerWithSeedInject:
+        if hasattr(config, "sample_std_dev"):
+            sample_std_dev = config.sample_std_dev
+        else:
+            sample_std_dev = get_default_mutation_std(config.bounds, target_level)
+        child_initializer = GaussianInitializerWithSeedInject(
+            seed=sprout_seed, std_dev=sample_std_dev, bounds=config.bounds
+        )
+    elif config.pop_initializer_class == LHSGlobalInitializer:
+        child_initializer = LHSGlobalInitializer(config.bounds, random_seed)
+    elif config.pop_initializer_class == SobolGlobalInitializer:
+        child_initializer = SobolGlobalInitializer(config.bounds, random_seed)
+    elif config.pop_initializer_class == InjectionInitializer:
+        if injected_population is None and sprout_seed is not None:
+            injected_population = [sprout_seed]
+        child_initializer = InjectionInitializer(injected_population, config.bounds)
```
This should be moved to another function?

WojtAcht replied:
Maybe it would be better to use match case: https://peps.python.org/pep-0622/ instead of if/elif

WojtAcht commented on test/test_deme_tree.py:
```diff
@@ -38,6 +38,7 @@ def test_deme_tree_properties(self):
         for leave in hms_tree.leaves:
             self.assertIsInstance(leave, CMADeme)
         for level, deme in hms_tree.all_demes:
+            print(level, deme)
```
TODO: remove

WojtAcht commented on pyhms/demes/initialize.py:
```diff
@@ -3,61 +3,90 @@
-    elif isinstance(config, SobolLevelConfig):
-        args["random_seed"] = random_seed
-        child = SobolDeme(**args)
+    elif isinstance(config, RandomLEvelConfig):
```
`RandomLEvelConfig` -> `RandomLevelConfig`

WojtAcht commented on pyhms/core/initializers.py:
```diff
@@ -0,0 +1,122 @@
+        else:
+            to_inject = [Individual(genome=ind.genome, problem=problem) for ind in self.injected_population]
+
+        if pop_size is None or pop_size >= len(self.injected_population):
```
I assume that sampled population has to be equal to `pop_size`. Why do we want to return `self.injected_population` if its size doesn't match `pop_size`?

Hubert-Guzowski replied:
I imagined a scenario, when we do have some evaluated individuals yet less, than required population size. In that case, we would want to use the injected population and sample some additional individuals.

WojtAcht replied:
ok, it makes sense

WojtAcht commented on pyhms/demes/initialize.py:
```diff
@@ -3,61 +3,100 @@
+def create_initializer(
+    config: BaseLevelConfig, sprout_seed: Individual | None, injected_population: list[Individual] | None
+) -> PopInitializer:
+    match config.pop_initializer_class.__qualname__:
```
Let's assume that I have a very custom initialization scheme that is problem specific. How can I use it without editing this function?

WojtAcht replied:
I implemented custom initialization for DistributionOptimization and created a draft PR: https://github.com/agh-a2s/pyhms/pull/79/files. The concept is straightforward: a new parameter in the configuration that specifies a function to create a new individual.

Hubert-Guzowski replied:
Yes, this would require editing a function to add a custom initialization scheme to the match-case. I think, that your new solution is better.

Hubert-Guzowski replied:
I will merge simple-initialization to this branch and then resolve conflicts with main

Hubert-Guzowski replied:
I went through it again and used the concept of including the initializer type in config. In my opinion it still had some issues: you still had to deduce some initialization properties during deme initialization based on some run context, which resulted in decision trees in SHADEDeme init.

Hubert-Guzowski commented:
I moved the initializer to config as suggested. Also, I decoupled the creation of initializer and providing the sampler with runtime context, which simplified the code. Now you can add a new, custom initializer and as long as it does not require any additional context-dependent information, you do not need to rewrite anything. If you needed it, there is a new function for that.

Conflicts with main are resolved. There was also a test_tree_format test file, which was not being detected by unittest discover, but was part of GitHub actions test suite, so I added a test class to it.

WojtAcht commented on pyhms/config.py:
```diff
@@ -32,14 +35,17 @@ def __init__(
         generations: int,
+        pop_initializer: type[PopInitializer] = UniformGlobalInitializer,
         sample_std_dev: float = 1.0,
+        initialize: Callable | None = None,
```
why do we need both `pop_initializer` and `initialize`?

WojtAcht commented on pyhms/demes/initialize.py:
```diff
@@ -1,67 +1,88 @@
+    sprout_seed: Individual | None,
+    injected_population: list[Individual] | None,
+) -> None:
+    context: dict[str, Any] = {}
```
I really don't like the use of context dictionaries in this code. It makes the code harder to understand and maintain.

Lack of Clarity: When you use a context dictionary, it's not immediately clear what fields are expected or required. There's no explicit contract or structure, so you have to dig through the code or documentation (if it even exists) to figure out what keys need to be provided. This ambiguity can easily lead to mistakes, like forgetting a key or passing the wrong data type, which might not be caught until runtime.

Increased Risk of Errors: Since the dictionary keys are often just strings, it's easy to make a typo or use an incorrect key, leading to subtle bugs that can be tricky to diagnose. For example, if the code expects a key like "seed_genome" but you accidentally pass "seedGenome", the error might not surface until much later, potentially in a different part of the program.

WojtAcht commented on pyhms/core/initializers.py:
```diff
@@ -0,0 +1,127 @@
+class PopInitializer(ABC):
+    def __init__(self, bounds: np.ndarray | None = None):
+        self._bounds = bounds
+        self._sampler: Callable = None
```
If I want to implement a completely new Initializer I have to do something like this:
```
class CustomInitializer(PopInitializer):
  def prepare_sampler(self, context: dict = None) -> None:
    self._sampler = my_custom_sampler
```
The problem is that I have no clue what's inside the `context`. I have to overwrite a private field of this class (it doesn't sound good). I also have to read the source code to understand that my sampler is a function called with a single argument `pop_size: int`.

WojtAcht replied:
Here you use `._sampler` but in `UniformGlobalInitializer` `.sampler` is used.

WojtAcht replied:
You also overwrite `sample_pop` (e.g. in `UniformGlobalInitializer`) which is not annotated as an abstract method.

WojtAcht replied:
The functionality is very simple, but you need to read many lines of code to understand what's going on. I think that it's overengineered. I don't like this approach.

WojtAcht replied:
Maybe a solution like this:
```
class PopulationInitializer:
    def __init__(self, sampler: Callable[[int], np.ndarray]):
        self.sampler = sampler

    def initialize(self, pop_size: int) -> np.ndarray:
        return self.sampler(pop_size)

def uniform_sampler(bounds: np.ndarray) -> Callable[[int], np.ndarray]:
    def sampler(pop_size: int) -> np.ndarray:
        return np.random.uniform(0, 1, (pop_size, len(bounds)))
    return sampler
```
would be simpler? we can generate multiple functions that create samplers (or a factory) and it's straightforward -- you have to just pass your function as an argument. The function takes one parameter `pop_size` and that's all.

WojtAcht commented on pyhms/core/initializers.py:
```diff
@@ -0,0 +1,127 @@
+        pass
+
+    def sample_pop(self, pop_size: int | None) -> np.ndarray:
+        return self._sampler(pop_size)
```
Why do you overwrite this method in `UniformGlobalInitializer`?

WojtAcht commented on pyhms/core/initializers.py:
```diff
@@ -0,0 +1,127 @@
+        return super().__call__(pop_size, problem) + [self.get_seed(problem)]
+
+
+class InjectionInitializer(SeededPopInitializer):
```
The code is complex and the functionality is very simple. It's confusing for me.

WojtAcht reviewed (changes requested):
I don't like this solution. It's really complex and I don't feel that it improves `pyhms`. For me it's overengineered.
